File: QHĐ - Dãy con tăng dài nhất (Chỉ có bản khó)/DPDSTT2.cpp
#pragma GCC optimize("O3")
#include <iostream>
#include <algorithm>

using namespace std;

int cnt, f[1000006], prime[80006], n, res = 0;
int **bit;

void eratos() {
    for(int i=2; i<=1000000; ++i) {
        if (f[i] == 0) prime[++cnt] = f[i] = i;
        int j = 1;
        while (j <= cnt && prime[j] <= f[i] && prime[j] * i <= 1000000) {
            f[prime[j] * i] = prime[j];
            ++j;
        }
    }
}

void init() {
    bit = new int*[cnt+1];
    for(int i=1; i<=cnt; ++i) {
        int m = 1000000 / prime[i];
        bit[i] = new int[m+1];
        for(int j=1; j<=m; ++j) bit[i][j] = 0;
    }
}

template<typename T>
bool fast_scan(T &num) {
    num = 0;
    register char c = getchar();
    while (c != '-' && (c < '0' || '9' < c)) {
        if (feof(stdin)) return false;
        c = getchar();
    }
    bool neg = false;
    if (c == '-') {
        neg = true;
        c = getchar();
    }
    for(; '0' <= c && c <= '9'; c = getchar()) num = (num << 1) + (num << 3) + (c - '0');
    if (neg) num = -num;
    return true;
}

void update(int block, int p, int val) {
    for(; p * prime[block] <= 1000000; p += (p & (-p))) bit[block][p] = max(bit[block][p], val);
}

int  query(int block, int p) {
    int ans = 0;
    for(; p>0; p -= (p & (-p))) ans = max(ans, bit[block][p]);
    return ans;
}

void update(int a, int p) {
    if (a <= 1) {
        res = max(res, 1);
        return;
    }
    int a1 = a;
    int best = 0;
    while (a > 1) {
        int tmp = f[a];
        int j = lower_bound(prime+1, prime+cnt+1, tmp) - prime, p = a1 / tmp;
        best = max(best, query(j, p-1) + 1);
        while (a % tmp == 0) a/= tmp;
    }
    a = a1; res = max(res, best);
    while (a > 1) {
        int tmp = f[a];
        int j = lower_bound(prime+1, prime+cnt+1, tmp) - prime, p = a1 / tmp;
        update(j, p, best);
        while (a % tmp == 0) a/= tmp;
    }
}

int main() {
    eratos();
    init();
    fast_scan(n);
    for(int i=1, a; i<=n; ++i) {
        fast_scan(a);
        update(a, i);
    }
    cout << res;

    return 0;
}
    

File: QHĐ - Dãy con tăng dài nhất (Chỉ có bản khó)/DPDSTT2.pas
uses        math;

type        arr_bit = array of longint;

var         f, prime, pos: array[0..1000006] of longint;
            bit: array[1..79006] of arr_bit;
            n, cnt, res, element: longint;

procedure   open_file;
begin
end;

procedure   close_file;
begin
end;

procedure   sieve;
var         i, j: longint;
begin
            cnt := 0;
            for i:=2 to 1000000 do
                begin
                    if f[i] = 0 then
                        begin
                            inc(cnt);
                            prime[cnt] := i;
                            pos[i] := cnt;
                            f[i] := i;
                        end;
                    j := 1;
                    while (j <= cnt) and (prime[j] <= f[i]) and (prime[j] * i <= 1000000) do
                        begin
                            f[prime[j]*i] := prime[j];
                            inc(j);
                        end;
                end;
end;

procedure   init;
var         i, j, tmp: longint;
begin
            res := 0;
            for i:=1 to cnt do
                begin
                    tmp := 1000000 div prime[i];
                    setLength(bit[i], tmp + 1);
                    for j:=0 to tmp do bit[i, j] := 0;
                end;
end;

procedure   update_bit(block, p, val: longint);
begin
            while p * prime[block] <= 1000000 do
                begin
                    bit[block, p] := max(val, bit[block, p]);
                    inc(p, p and (-p));
                end;
end;

function    query_bit(block, p: longint): longint;
var         ans : longint = 0;
begin
            while p > 0 do
                begin
                    ans := max(ans, bit[block, p]);
                    dec(p, p and (-p));
                end;
            exit(ans);
end;

procedure   update(a: longint);
var         a1, tmp, best, p: longint;
begin
            if a <= 1 then
                begin
                    res := max(res, 1);
                    exit;
                end;
            best := 0;
            a1 := a;
            while a > 1 do
                begin
                    tmp := f[a];
                    p := a1 div tmp;
                    best := max(best, query_bit(pos[tmp], p-1) + 1);
                    while a mod tmp = 0 do a := a div tmp;
                end;
            a := a1;
            res := max(res, best);
            while a > 1 do
                begin
                    tmp := f[a];
                    p := a1 div tmp;
                    update_bit(pos[tmp], p, best);
                    while a mod tmp = 0 do a := a div tmp;
                end;
end;

begin
            open_file;

            sieve;
            init;
            readln(n);
            while n > 0 do
                begin
                    read(element);
                    update(element);
                    dec(n);
                end;
            write(res);

            close_file;
end.
    
